### DISCUSSION TITLE
Customizing TinaCMS UI's styling

### DISCUSSION THREAD

**Discussion by jamespohalloran** [tinacms/tinacms#3412] [Ideas]
Copying over some thoughts from @Perni1984 around customizing TinaCMS's styles:

> our concrete use case is to overwrite the css styles in slate so that the WYSIWYG Editor headline styles better match the actual headline styles of the real site.

> I digged a bit into the code and it seems that there is no way currently to hook into the vite build process. Any way to add a custom config or a custom vite plugin would be highly appreciated.

**Comment by jeffsee55** [comment]
We can expose the function we use to render components, it's pretty simple and we could wrap it up to be a bit easier, it'd work almost identically to `<TinaMarkdown>` except with a few extra props. And you'd supply it as part of the field definition:

```ts
{
  type: 'rich-text',
  ui: {
    components: {
      h1: ({ attributes, element, className, ...props }) => (
        <h1
          className={`${className} my-custom-class`}
          {...attributes}
          {...props}
        />),
      },
      h2: ...
    }
  }
}
```

See the actual code for this [here](https://github.com/tinacms/tinacms/blob/431d7398039e9d84d5bcde2bec722f0791d17166/packages/@tinacms/toolkit/src/packages/fields/plugins/MdxFieldPlugin/plate/plugins/ui/components.tsx#L54-L286)

**Reply by Perni1984** [reply]
@jeffsee55 that would be a good start, but actually we are currently styling two separate things with our approach - the Headlines and Paragraph styles in the Rich Text editor and also the ones in the Dropdown component in the toolbar.

Moreover we're using [vanilla-extract](https://vanilla-extract.style/) as our css library, so I guess for us being able to use it we would need to add the [vanilla-extract/vite-plugin](https://vanilla-extract.style/documentation/integrations/vite/) to the vite configuration.

**Reply by ghost** [reply]
Hi everyone.

I don't know if you know, there is an open source editor that has this feature that you are talking about here. The name of this text editor is called hope.js - please see this `https://github.com/poef/hope`

*idea-before*
```
{
  type: 'rich-text',
  ui: {
    components: {
      h1: ({ attributes, element, className, ...props }) => (
        <h1
          className={`${className} my-custom-class`}
          {...attributes}
          {...props}
        />),
      },
      h2: ...
    }
  }
}
```

*idea-after*
```
// components
0-44:'rich-text'
43-44:ui#my-custom-class
46-80:h1#my-custom-class
46-80:h2#my-custom-class
```

*My idea based on hope.js and tina-cms would be something like this. so I'd like to know what you all think of this idea, of a component-chained block UI.*
